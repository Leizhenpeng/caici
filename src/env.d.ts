



interface ImportMetaEnv extends Readonly<Record<string, string>> {
  readonly VITE_MODE_NAME: string,
  readonly VITE_SOCKET_URL: string,
}
interface ImportMeta {
  readonly env: ImportMetaEnv
}
