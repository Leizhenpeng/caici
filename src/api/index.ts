export * from './axios'
export * from './test'
export * from './user'
export * from './together'
