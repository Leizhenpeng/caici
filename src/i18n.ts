import hans from './locales/zh-cn.json'
import hant from './locales/zh-tw.json'

const lang = window.navigator.language?.toLowerCase() || ''
export const preferTraditional = lang.includes('hant') || lang.includes('tw') || lang.includes('hk')
export const preferZhuyin = lang.includes('tw')

export const locale = useStorage<'hans' | 'hant'>('handle-locale', preferTraditional ? 'hant' : 'hans')

export function t(key: keyof typeof hans, ...t: any[]): string {
  const str: string = (locale.value === 'hant' ? (hant as any)[key] : null) || hans[key]
  return str.replace(/\{(\d+)\}/g, (_, i) => t[i])
}
