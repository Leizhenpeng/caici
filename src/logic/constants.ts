export const WORD_LENGTH = 4
export const TRIES_LIMIT = 10
export const START_DATE = new Date(2022, 0, 0)
export const RANDOM_SEED = 'handle'
export const DAYS_PLAY_BACK = 3
export const DEPLOY_HOST = 'handle.antfu.me'
export const NETLIFY_FUNCTION_HOST = '/.netlify/functions'
