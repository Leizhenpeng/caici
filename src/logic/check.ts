import { WORD_LENGTH } from './constants'
import { getIdiom } from './idioms'

export function filterNonChineseChars(input: string) {
  return Array.from(input)
    .filter(i => /\p{Script=Han}/u.test(i))
    .slice(0, WORD_LENGTH)
    .join('')
}

export function checkValidIdiom(word: string, strict = false) {
  if (!strict)
    return true
  return !!getIdiom(word)
}
