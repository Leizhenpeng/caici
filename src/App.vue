<script setup lang="ts">
import '~/init'
import { answer, dayNo, daySince, isDev } from '~/state'
import { colorblind } from '~/storage'
import { DAYS_PLAY_BACK } from '~/logic/constants'

const { height } = useWindowSize()

watchEffect(() => {
  document.documentElement.style.setProperty('--vh', `${height.value / 100}px`)
})
</script>

<template>
  <main font-sans text="center gray-700 dark:gray-300" :class="{ colorblind }">
    <NotTodayBanner v-if="dayNo < daySince" />
    <Navbar />
    <div p="4">
      <NoQuizToday v-if="!answer.word" />
      <NoFuturePlay v-else-if="dayNo > daySince && !isDev" />
      <NoPastPlay v-else-if="daySince - dayNo > DAYS_PLAY_BACK && !isDev" />
      <Play v-else />
    </div>
    <ModalsLayer />
    <Confetti />
  </main>
</template>
