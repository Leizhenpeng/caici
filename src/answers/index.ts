import { getHint } from '../logic'
import { answers } from './list'

export function getAnswerOfDay(day: number) {
  const [word = '', hint = ''] = answers[day] || []
  return {
    word,
    hint: hint || getHint(word),
  }
}
