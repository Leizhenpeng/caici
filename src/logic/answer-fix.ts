import { TRIES_LIMIT } from './constants'
import { historyMeta } from '~/storage'
import { getAnswerOfDay } from '~/answers'

export function tryFixAnswer(day: number) {
  const meta = historyMeta.value[day]
  const answer = getAnswerOfDay(day)
  if (!meta)
    return
  if (!meta.answer && !meta.failed && !meta.passed)
    return

  const tries = meta.tries || []
  const index = tries.indexOf(answer.word)
  if (index < 0 || index >= tries.length - 1)
    return

  const newTries = tries.slice(0, index + 1)
  meta.tries = newTries
  if (index <= TRIES_LIMIT) {
    meta.passed = true
    meta.failed = false
    meta.answer = false
  }
  meta.duration = (meta.duration || 0) * newTries.length / tries.length
}
