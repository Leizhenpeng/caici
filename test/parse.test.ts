import { expect, it } from 'vitest'
import { parseChar, parseWord, testAnswer } from '../src/logic'

it('parseChar', () => {
  expect(parseChar('虎')).toMatchSnapshot()
  expect(parseChar('微')).toMatchSnapshot()
  expect(parseChar('姜')).toMatchSnapshot()
  expect(parseChar('寿')).toMatchSnapshot()
  expect(parseChar('翁')).toMatchSnapshot()
})

it('testAnswer', () => {
  expect(testAnswer(parseWord('安'), parseWord('安'))).toMatchSnapshot()
  expect(testAnswer(parseWord('帆'), parseWord('安'))).toMatchSnapshot()
})
