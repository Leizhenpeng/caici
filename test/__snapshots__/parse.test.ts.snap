// Vitest Snapshot v1

exports[`parseChar 1`] = `
{
  "_1": "h",
  "_2": "u",
  "_3": undefined,
  "char": "虎",
  "parts": [
    "h",
    "u",
  ],
  "tone": 3,
  "yin": "hu",
}
`;

exports[`parseChar 2`] = `
{
  "_1": "w",
  "_2": "ei",
  "_3": undefined,
  "char": "微",
  "parts": [
    "w",
    "ei",
  ],
  "tone": 1,
  "yin": "wei",
}
`;

exports[`parseChar 3`] = `
{
  "_1": "j",
  "_2": "iang",
  "_3": undefined,
  "char": "姜",
  "parts": [
    "j",
    "iang",
  ],
  "tone": 1,
  "yin": "jiang",
}
`;

exports[`parseChar 4`] = `
{
  "_1": "sh",
  "_2": "ou",
  "_3": undefined,
  "char": "寿",
  "parts": [
    "sh",
    "ou",
  ],
  "tone": 4,
  "yin": "shou",
}
`;

exports[`parseChar 5`] = `
{
  "_1": "w",
  "_2": "eng",
  "_3": undefined,
  "char": "翁",
  "parts": [
    "w",
    "eng",
  ],
  "tone": 1,
  "yin": "weng",
}
`;

exports[`testAnswer 1`] = `
[
  {
    "_1": "exact",
    "_2": "exact",
    "_3": "exact",
    "char": "exact",
    "tone": "exact",
  },
]
`;

exports[`testAnswer 2`] = `
[
  {
    "_1": "none",
    "_2": "exact",
    "_3": "exact",
    "char": "none",
    "tone": "exact",
  },
]
`;
